skglm/solvers/prox_newton.py

```diff
@@ -0,0 +1,338 @@
+            old_w_idx = w_ws[idx]
+            stepsize = 1 / lipschitz[idx]
+
+            w_ws[idx] = penalty.prox_1d(
+                old_w_idx - stepsize * cached_grads[idx], stepsize, j)
+
+            if w_ws[idx] != old_w_idx:
+                X_delta_w_ws += (w_ws[idx] - old_w_idx) * X[:, j]
+
+        if cd_iter % 5 == 0:
+            opt = penalty.subdiff_distance(w_ws, cached_grads, ws)
+            if np.max(opt) <= tol:
+                break
+
+    # descent direction
+    return w_ws - w_epoch[ws], X_delta_w_ws
+
+
+# sparse version of _compute_descent_direction
+@njit
+def _descent_direction_s(X_data, X_indptr, X_indices, y, w_epoch,
+                         Xw_epoch, grad_ws, datafit, penalty, ws, tol):
+    raw_hess = datafit.raw_hessian(y, Xw_epoch)
+
+    lipschitz = np.zeros(len(ws))
+    for idx, j in enumerate(ws):
+        # equivalent to: lipschitz[idx] += raw_hess * X[:, j] ** 2
+        lipschitz[idx] = _sparse_squared_weighted_norm(X_data, X_indptr, X_indices,
+                                                       j, raw_hess)
+
+    cached_grads = np.zeros(len(ws))
+    X_delta_w_ws = np.zeros(len(y))
+    w_ws = w_epoch[ws]
+
+    for cd_iter in range(MAX_CD_ITER):
+        for idx, j in enumerate(ws):
+            # skip when X[:, j] == 0
+            if lipschitz[idx] == 0:
+                continue
+
+            cached_grads[idx] = grad_ws[idx]
+            # equivalent to cached_grads[idx] += X[:, j] @ (raw_hess * X_delta_w_ws)
+            cached_grads[idx] += _sparse_weighted_dot(X_data, X_indptr, X_indices, j,
+                                                      X_delta_w_ws, raw_hess)
+
+            old_w_idx = w_ws[idx]
+            stepsize = 1 / lipschitz[idx]
+
+            w_ws[idx] = penalty.prox_1d(
+                old_w_idx - stepsize * cached_grads[idx], stepsize, j)
+
+            if w_ws[idx] != old_w_idx:
+                _update_X_delta_w(X_data, X_indptr, X_indices, X_delta_w_ws,
+                                  w_ws[idx] - old_w_idx, j)
+
+        if cd_iter % 5 == 0:
+            opt = penalty.subdiff_distance(w_ws, cached_grads, ws)
+            if np.max(opt) <= tol:
+                break
+
+    # descent direction
+    return w_ws - w_epoch[ws], X_delta_w_ws
+
+
+@njit
+def _backtrack_line_search(X, y, w, Xw, datafit, penalty, delta_w_ws,
+                           X_delta_w_ws, ws):
+    # inplace update of w and Xw
+    # return grad_ws of the last w and Xw
+    step, prev_step = 1., 0.
+    epoch_penalty_val = penalty.value(w[ws])
+
+    for backtrack_iter in range(MAX_BACKTRACK_ITER):
+        stop_crit = -epoch_penalty_val
+        w[ws] += (step - prev_step) * delta_w_ws
+        Xw += (step - prev_step) * X_delta_w_ws
+
+        grad_ws = _construct_grad(X, y, w, Xw, datafit, ws)
+        stop_crit += step * grad_ws @ delta_w_ws
+        stop_crit += penalty.value(w[ws])
+
+        if stop_crit < 0:
```

Comment:
Finally, since we are computing the descent direction approximately, is it possible that the backtracking line search does not provide a better objective? In other words what is happening if `MAX_BACKTRACK_ITER` is attained? Should we handle this corner case? any idea @mathurinm @Klopfe ?

Reply:
In addition, I think we might need to think about to scope of the penalties taken into account, is it possible to use this type of criterion for non-convex penalties? At least maybe we could add a warning in the doc.

Reply:
I have put a TODO. We can see what happens with non convex penalties experimentally later if that is OK